control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -3743,17 +3745,66 @@ func (r *HostedControlPlaneReconciler) reconcileOperatorLifecycleManager(ctx con
 				return fmt.Errorf("failed to reconcile catalogs: %w", err)
 			}
 
-			catalogsImageStream := manifests.CatalogsImageStream(hcp.Namespace)
-			if !overrideImages {
-				isImageRegistryOverrides := util.ConvertImageRegistryOverrideStringToMap(p.OLMCatalogsISRegistryOverridesAnnotation)
-				if _, err := createOrUpdate(ctx, r, catalogsImageStream, func() error {
-					return olm.ReconcileCatalogsImageStream(catalogsImageStream, p.OwnerRef, isImageRegistryOverrides)
-				}); err != nil {
-					return fmt.Errorf("failed to reconcile catalogs image stream: %w", err)
+			isImageRegistryOverrides := util.ConvertImageRegistryOverrideStringToMap(p.OLMCatalogsISRegistryOverridesAnnotation)
+
+			if r.ManagementClusterCapabilities.Has(capabilities.CapabilityImageStream) {
+				catalogsImageStream := manifests.CatalogsImageStream(hcp.Namespace)
+				if !overrideImages {
+					if _, err := createOrUpdate(ctx, r, catalogsImageStream, func() error {
+						return olm.ReconcileCatalogsImageStream(catalogsImageStream, p.OwnerRef, isImageRegistryOverrides)
+					}); err != nil {
+						return fmt.Errorf("failed to reconcile catalogs image stream: %w", err)
+					}
+				} else {
+					if _, err := util.DeleteIfNeeded(ctx, r, catalogsImageStream); err != nil {
+						return fmt.Errorf("failed to remove OLM Catalog ImageStream: %w", err)
+					}
 				}
-			} else if r.ManagementClusterCapabilities.Has(capabilities.CapabilityImageStream) {
-				if _, err := util.DeleteIfNeeded(ctx, r, catalogsImageStream); err != nil {
-					return fmt.Errorf("failed to remove OLM Catalog ImageStream: %w", err)
+			} else {
+				if !overrideImages {
+					pullSecret := common.PullSecret(hcp.Namespace)
+					if err := r.Get(ctx, client.ObjectKeyFromObject(pullSecret), pullSecret); err != nil {
+						return fmt.Errorf("failed to get pull secret for namespace %s: %w", hcp.Namespace, err)
+					}
+
+					for name, catalog := range olm.CatalogToImage {
+						dockerImageRef, err := reference.Parse(catalog)
+						if err != nil {
+							return fmt.Errorf("failed to parse catalog image %s: %w", catalog, err)
+						}
+
+						if len(isImageRegistryOverrides) > 0 {
+							for registrySource, registryDest := range isImageRegistryOverrides {
+								if strings.Contains(dockerImageRef.Exact(), registrySource) {
```

Comment:
why not just compare the registry ? 
if imageRef.Registry == registrySource

Reply:
thats what I had done originally but noticed that what we do in the code above is better because with the override we can point to a different registry, registry namespace and registry repo so it gives us more control.